test/distributed/fsdp/test_fsdp_summon_full_params.py

```diff
@@ -0,0 +1,176 @@
+    def test_summon_full_param_no_writeback(self):
+        model = FSDP(nn.Linear(4, 4).cuda(self.rank))
+        # set the value
+        p = list(model.parameters())[0]
+        with torch.no_grad():
+            # This sets the local shard value
+            p[0] = self.rank + 2
+
+        with model._summon_full_params(writeback=False):
+            with torch.no_grad():
+                p.copy_(torch.zeros_like(p))
+
+        self.assertEqual(p[0], self.rank + 2)
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_with_writeback(self):
+        model = FSDP(nn.Linear(3, 3).cuda(self.rank))
+        # set the value
+        p = list(model.parameters())[0]
+        with torch.no_grad():
+            # This sets the local shard value
+            p[0] = self.rank + 2
+
+        with torch.no_grad():
+            # writeback = True is the default
+            with model._summon_full_params():
+                p.copy_(torch.zeros_like(p))
+
+        self.assertEqual(p[0], 0)
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_shard_value(self):
+
+        raw_model = nn.Linear(10, 11)
+        raw_model_size = self.get_model_param_count(raw_model)
+        expected_shard_size = self.get_expected_sharded_size(raw_model_size)
+
+        model = FSDP(raw_model.cuda(self.rank))
+        self.assertEqual(expected_shard_size, self.get_model_param_count(model))
+
+        # we're assuming a single flatenned param
+        self.assertEqual(1, len(list(model.parameters())))
+
+        my_shard = torch.clone(next(model.parameters()))
+
+        with model._summon_full_params():
+            self.assertEqual(raw_model_size, self.get_model_param_count(model))
+            all_shards = next(model.parameters())
+            my_slice = torch.chunk(all_shards, self.world_size)[self.rank]
+
+            # shards are padded but the full_param tensor is not
+            self.assertTrue(torch.equal(my_shard[0: my_slice.numel()], my_slice))
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_recursive(self):
+        model = FSDP(nn.Sequential(
+            FSDP(nn.Linear(5, 5, bias=False)),
+            nn.Linear(5, 3, bias=False)
+        )).cuda(self.rank)
+
+        global_inner_numel = self.get_model_param_count(nn.Linear(5, 5, bias=False))
+        global_outer_numel = self.get_model_param_count(nn.Linear(5, 3, bias=False))
+
+        shard_inner_numel = int(math.ceil(global_inner_numel / self.world_size))
+        shard_outer_numel = int(math.ceil(global_outer_numel / self.world_size))
+
+        outer_param = model.get_parameter("_fsdp_wrapped_module.flat_param")
+        inner_param = model.get_parameter("_fsdp_wrapped_module._fpw_module.0._fsdp_wrapped_module.flat_param")
+        self.assertEqual(shard_outer_numel, outer_param.numel())
+        self.assertEqual(shard_inner_numel, inner_param.numel())
+
+        # non recursive, should only summon outer
+        with model._summon_full_params(recurse=False):
```

Comment:
nit: can reduce duplication a bit:

```
for recurse in [true, false]:
with summon_full_params(recurse):
     self.assertEqual(global_outer_numel, outer_param.numel())
     self.assertEqual(shard_inner_numel, inner_param.numel())
```

Reply:
I changed the test to use `@parametrize`